core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1377,7 +1379,13 @@ private PlanWithProperties arbitraryDistributeUnion(
                 // instead of "arbitraryPartition".
                 return new PlanWithProperties(node.replaceChildren(partitionedChildren));
             }
-            // Trino currently cannot execute stage that has multiple table scans, so in that case
+            // all children are source partitioned - no intermediate RemoteExchange
+            if (countSources(partitionedChildren) == partitionedChildren.size()
```

Comment:
I think this predicate is wrong, see
```
                              union_all => countSources = 2 == 2
                                |
           ----------------------------------
           |                                |
      union_all                           aggregation(final)[hash_by=x]
        |                                   |
---------------------                     exchange(hash)
|                   |
table_scan(a) table_Scan(b)
```

it wouldn't put `FIXED_ARBITRARY_DISTRIBUTION` exchange, but it should do so above `aggregation(final)[hash_by=x]`

You need to actually look at the properties of the children. Check if each child is `child.getProperties().getNodePartitioning().get().getHandle() == SOURCE_DISTRIBUTION;`

Then you also don't need `hasPartitionedConnectorSource` check.

Please add a test